plat/kvm/x86/traps.c

```diff
@@ -27,135 +27,195 @@
 /* Declare the traps used only by this platform: */
 DECLARE_TRAP_EC(nmi,           "NMI",                  NULL)
 DECLARE_TRAP_EC(double_fault,  "double fault",         NULL)
 DECLARE_TRAP_EC(virt_error,    "virtualization error", NULL)
 
+static struct seg_gate_desc64 cpu_idt[IDT_NUM_ENTRIES] __align(8);
+static struct desc_table_ptr64 idtptr;
 
-static struct seg_gate_desc64 cpu_idt[IDT_NUM_ENTRIES] __align64b;
-
-static void idt_fillgate(unsigned int num, void *fun, unsigned int ist)
+static inline void idt_fillgate(unsigned int num, void *fun, unsigned int ist)
 {
 	struct seg_gate_desc64 *desc = &cpu_idt[num];
 
 	/*
 	 * All gates are interrupt gates, all handlers run with interrupts off.
 	 */
-	desc->offset_hi = (__u64) fun >> 16;
-	desc->offset_lo = (__u64) fun & 0xffff;
-	desc->selector = GDT_DESC_OFFSET(GDT_DESC_CODE);
-	desc->ist = ist;
-	desc->type = 14; /* == 0b1110 */
-	desc->dpl = 0;
-	desc->p = 1;
+	desc->offset_hi	= (__u64) fun >> 16;
+	desc->offset_lo	= (__u64) fun & 0xffff;
+	desc->selector	= GDT_DESC_OFFSET(GDT_DESC_CODE);
+	desc->ist	= ist;
+	desc->type	= 14; /* == 0b1110 */
+	desc->dpl	= 0;
+	desc->p		= 1;
 }
 
-volatile struct desc_table_ptr64 idtptr;
-
 static void idt_init(void)
 {
+	if (idtptr.limit != 0) {
+		__asm__ __volatile__("lidt %0" :: "m" (idtptr));
+		return;
+	}
+
 	/*
-	 * Load trap vectors. All traps run on IST2 (cpu_trap_stack), except for
-	 * the exceptions.
+	 * Load trap vectors. All traps run on the current stack, except
+	 * for critical and debug exceptions.
+	 *
+	 * NOTE: Nested IRQs/exceptions must not use IST as the CPU switches
+	 * to the configured stack pointer irrespective of the fact that it may
+	 * already run on the same stack. For example, if we would cause a
+	 * page fault in the debug trap and both would use the same IST entry,
+	 * the page fault handler corrupts the stack (of the debug trap) and
+	 * the system eventually crashes when returning from the page fault
+	 * handler.
 	 */
-#define FILL_TRAP_GATE(name, ist) extern void cpu_trap_##name(void); \
+#define FILL_TRAP_GATE(name, ist)					\
+	extern void cpu_trap_##name(void);				\
 	idt_fillgate(TRAP_##name, ASM_TRAP_SYM(name), ist)
-	FILL_TRAP_GATE(divide_error,    2);
-	FILL_TRAP_GATE(debug,           2);
-	FILL_TRAP_GATE(nmi,             3); /* #NMI runs on IST3 (cpu_nmi_stack) */
-	FILL_TRAP_GATE(int3,            2);
-	FILL_TRAP_GATE(overflow,        2);
-	FILL_TRAP_GATE(bounds,          2);
-	FILL_TRAP_GATE(invalid_op,      2);
-	FILL_TRAP_GATE(no_device,       2);
-	FILL_TRAP_GATE(double_fault,    3); /* #DF runs on IST3 (cpu_nmi_stack) */
-
-	FILL_TRAP_GATE(invalid_tss,     2);
-	FILL_TRAP_GATE(no_segment,      2);
-	FILL_TRAP_GATE(stack_error,     2);
-	FILL_TRAP_GATE(gp_fault,        2);
-	FILL_TRAP_GATE(page_fault,      2);
-
-	FILL_TRAP_GATE(coproc_error,    2);
-	FILL_TRAP_GATE(alignment_check, 2);
-	FILL_TRAP_GATE(machine_check,   2);
-	FILL_TRAP_GATE(simd_error,      2);
-	FILL_TRAP_GATE(virt_error,      2);
+
```

Comment:
The trap function assignment and pointer assignment should be separated, because `traps_init()` is called by `lcpu_arch_init()` and `lcpu_arch_init()` is called separately for each CPUs. This will result in unnecessary calls to `FILL_TRAP_GATE()` and `FILL_IRQ_GATE()`. 
You can write two separate functions :
`init_trap_vectors()` which will be called only once.
`init_idtptr()` which will initialize idtptr for each CPU.

Reply:
This is not the case, because there is a check in lines [169 ff](https://github.com/unikraft/unikraft/blob/8096a7885862a52ce721affa9097ee363392419e/plat/kvm/x86/traps.c#L169) that is true for all secondary cores and which leads to an early exit that only sets the IDT based on the existing table.

Reply:
Yeah, you are right

Reply:
Actually, I had the same thought when I was updating the functions but I wanted to keep all three functions looking and working the same (i.e., do all necessary init and activate). Also I had the following thoughts. If we separate initialization and activation, we have three options to do this:

1. Do a similar check in `traps_init()` based on the `lcpu` and call initialization only for BSP
2. Export initialization function in addition to `traps_init()` and call it in the common x86 code in `lcpu_arch_init()`, and do the check for the BSP there
3. Export initialization function and call it in `setup.c` before `lcpu_init()`

The first option would not bring any real benefit because we are still doing the check and now this function is the only one of the three which exposes the necessity for `traps_init` to differentiate between BSP and APs. The second option is the same only uglier. 

The third option is the only one which would actually remove the check altogether and make sense from an optimization and design point of view. The somewhat ugly aspect here is, however, that the `lcpu` is not yet initialized (not that bad because we don't need it) and I felt that doing a separate `trap_vectors_init()` before the `lcpu_init()` unnecessarily exposes a part of the CPU initialization (i.e., what `lcpu_init()` is supposed to do).

There is also an alternative solution:
4. Initialize the IDT statically at compile-time, thereby removing the run-time initialization altogether. This would avoid all the aforementioned problems and improve boot time! I did a quick trial to do this with assembler. The problem here is that you cannot do mask and shift operations on symbols and it is always a pain to build a loop with an iterator variable because you need to nest macros for this. So this only works with a more powerful python script that is invoked during the build and I thought, I leave this for the future, because it also affects the ability to do ASLR (symbol addresses).

What are your thoughts on all of this?

Reply:
@Sairajkodilkar ping

Reply:
Hi @marcrittinghaus, you should go with method 3, where you call the traps initialisation function in setup.c and idtptr initialisation in `lcpu_init()`. 
I think this does does not expose the part of CPU init because we are just initialising trap gates in setup.c. 
`lcpu_init()` does the actual initialisation by assigning idtptr

Reply:
@Sairajkodilkar Ok, I will update the PR accordingly.